### DISCUSSION TITLE
How to embed a custom win32 window to the WinUI3 application

### DISCUSSION THREAD

**Discussion by evhenii-komar** [microsoft/microsoft-ui-xaml#9912] [Q&A]
Hi! I've added an extra win32 window to my WinUi3 application. It works as expected on Windows 10, but I had a problem with Windows 11. My window is visible but doesn't react on the mouse or receive any messages. All keyboard input goes to the window whose class is `Microsoft.UI.Content.DesktopChildSiteBridge` and I found that mouse input goes to the window whose class is  `MITMessageOnlyWindowClass`. The last one is not in the WinUi3 application window hierarchy and even Microsoft Spy++ doesn't show it in the desktop window hierarchy.
I saw that WebView2 embeds its extra window as a child of the `Microsoft.UI.Content.DesktopChildSiteBridge` window however I didn't find any information on how to do it.
I found this topic - [How to embed a child window into the window in WinUI3](https://github.com/castorix/WinUI3_SwapChainPanel_Layered/issues/3). It discusses a similar situation, but if I understood right the code in the project of the guy who answered the topic author I should redirect application windows messages to my extra window with Win API functions. It was in 2023. I'm wondering if there is a more elegant solution at this point.

**Comment by castorix** [comment]
> Hi! I've added an extra win32 window to my WinUi3 application. It works as expected on Windows 10, but I had a problem with Windows 11. My window is visible but doesn't react on the mouse or receive any messages.

Do you mean that in the _WinUI3_DesktopDuplication_ test, you cannot type keys in the Win32 control on Windows 11 ?
WebView2 also uses _ContentExternalOutputLink_, but it is still marked experimental and on some tests I had done, I could only "transform" a Win32 control into XAML, but without input messages either (maybe I forgot some things...)

**Comment by softworkz** [comment]
> All keyboard input goes to the window whose class is `Microsoft.UI.Content.DesktopChildSiteBridge` and I found that mouse input goes to the window whose class is `MITMessageOnlyWindowClass`. The last one is not in the WinUi3 application window hierarchy and even Microsoft Spy++ doesn't show it in the desktop window hierarchy.

The `MITMessageOnlyWindowClass` window is a so-called "message-only" window. Those windows cannot be enumerated (that's why Spy++ doesn't show it).  These windows can still be "found" by using `FindWindowEx` with `HWND_MESSAGE` as parent window.
Yet, that doesn't help much in this case.

> I saw that WebView2 embeds its extra window as a child of the Microsoft.UI.Content.DesktopChildSiteBridge window however I didn't find any information on how to do it.

This window

![image](https://github.com/user-attachments/assets/d66f7917-a21b-458e-854e-976506149539)

doesn't convey any input messages. It's probably there for associating the WebView2 with the window where it's been used. 

But the actual input is sent to the WebView2 via methods on the [CoreWebView2CompositionController](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2compositioncontroller) like [SendMouseInput](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2compositioncontroller?view=webview2-winrt-1.0.2592.51#sendmouseinput) and [SendPointerInput](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2compositioncontroller?view=webview2-winrt-1.0.2592.51#sendpointerinput). 

Further, that `Chrome_WidgetWin_0` window does not render the browser content (this is going through some insane interop hack, caused by the very unfortunate limitations of the app-local composition architecture in WinUI3).

So all-in-all the WebView2 cannot serve as an example for embedding win32 windows in a WinUI3 app.

**Comment by softworkz** [comment]
I had some success in enabling input for an embedded hWnd inside a WinUI3 window by using an undocumented function in `Microsoft.InputStateManager.dll`. This allows to disable input processing on the XAML side and lets your embedded hWnd window receive keyboard and mouse events. The problems with this method are:

- there's no way to get pointer events
  (I've gone great lengths trying all win32 API tricks I could think of and beyond...)
- After doing that, your XAML doesn't get input events anymore
  I didn't need it. Further reverse-engineering of Microsoft.InputStateManager would probably reveal a way to re-enable input for the XAML side

**Comment by softworkz** [comment]
Please see my post here for an explanation why it doesn't work and my proposal to remedy: https://github.com/microsoft/microsoft-ui-xaml/issues/10050

**Comment by manoj-sunagar** [comment]
can anyone please take a look at this issue : https://github.com/microsoft/microsoft-ui-xaml/issues/10319
